pkg/relayer/relayer.go

```diff
@@ -0,0 +1,49 @@
+package relayer
+
+import (
+	"context"
+
+	"cosmossdk.io/depinject"
+)
+
+type RelayerOption func(*Relayer)
+
+// Relayer is the main struct that encapsulates the relayer's responsibilities.
+// It starts and stops the RelayerProxy and provide the served relays observable to them miner.
+type Relayer struct {
+	relayerProxy RelayerProxy
+	miner        Miner
+}
+
+// NewRelayer creates a new Relayer instance with the given dependencies.
+// It injects the dependencies into the Relayer instance and returns it.
+func NewRelayer(
+	deps depinject.Config,
+	opts ...RelayerOption,
+) (*Relayer, error) {
+	rel := &Relayer{}
+
+	if err := depinject.Inject(deps, &rel.relayerProxy, &rel.miner); err != nil {
+		return nil, err
+	}
+
+	for _, opt := range opts {
+		opt(rel)
+	}
+
+	return rel, nil
+}
+
+// Start provides the miner with the served relays observable and starts the relayer proxy.
+// This method is blocking while the relayer proxy is running and returns when Stop is called
+// or when the relayer proxy fails to start.
+func (rel *Relayer) Start(ctx context.Context) error {
+	rel.miner.MineRelays(ctx, rel.relayerProxy.ServedRelays())
```

Comment:
Not clear if this is a background go routine or not. I think it means we either have to add comments and/or rename downstream functions.

maybe `rel.miner.StartMining`?

Reply:
Not a background go routine.
Add comments: Yes.
Rename to `StartMining[Relays]`: Yes
:+1:

Reply:
`#MineRelays()` will not block:

```
// MineRelays kicks off relay mining by mapping the servedRelays ovservable through
// a pipeline which hashes the relay, checks if it's above the mining difficulty,
// adds it to the session tree, and then maps any errors to a  new observable.
// It also starts the claim and proof pipelines which are subsequently driven by
// mapping over RelayerSessionsManager's SessionsToClaim return observable.
// It does not block as map operations run in their own goroutines.
```